### PR TITLE
[Validator] Use PHP attributes when creating custom validation constraints

### PR DESCRIPTION
**Pull request by yoannrenard**
Describe how to use PHP attributes when creating custom validation constraints

### CODE REVIEW

File: validation/custom_constraint.rst [hunk 1]
```diff
@@ -12,20 +12,37 @@ alphanumeric characters.
...
+        }
+
+    .. code-block:: php-attributes
+
+        // src/Validator/ContainsAlphanumeric.php
+        namespace App\Validator;
+
+        use Symfony\Component\Validator\Constraint;
+
+        #[\Attribute()]
```
Comment (OskarStark) [inline validation/custom_constraint.rst]: @derrabus do we recommend a leading slash and the brackets?

File: validation/custom_constraint.rst [hunk 2]
```diff
@@ -128,6 +145,25 @@ You can use custom validators like the ones provided by Symfony itself:
...
+
+        use App\Validator as AcmeAssert;
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class AcmeEntity
+        {
+            // ...
+
+            #[Assert\NotBlank()]
+            #[AcmeAssert\ContainsAlphanumeric()]
```
Comment (OskarStark) [inline validation/custom_constraint.rst]:
```suggestion
            #[Assert\NotBlank]
            #[AcmeAssert\ContainsAlphanumeric]
```

File: validation/custom_constraint.rst [hunk 3]
```diff
@@ -241,6 +277,14 @@ not to the property:
             // ...
         }
 
+    .. code-block:: php-attributes
+
+        #[AcmeAssert\ProtocolClass()]
```
Comment (OskarStark) [inline validation/custom_constraint.rst]:
```suggestion
        #[AcmeAssert\ProtocolClass]
```

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: validation/custom_constraint.rst (hunk 1 above)
Comment (yoannrenard) [reply validation/custom_constraint.rst]: @OskarStark 
I got rid of the brackets like it is in the php annotations snippets.  
Though the leading slash looks required to me as it doesn't work without.

Comment (derrabus) [reply validation/custom_constraint.rst]: @OskarStark Symfony's CS policy is that we don't import classes from the global namespace, so the leading backslash is correct.

**Comment by derrabus** [comment]
The PR should target 5.2 and we should add a note that attribute constraints have been added with Symfony 5.2.

**Comment by yoannrenard** [comment]
I've added a small note inspired by the one made for [creating routes as attributes](https://symfony.com/doc/current/routing.html#creating-routes-as-attributes-or-annotations) as you requested @derrabus and changed the PR target to 5.2

**Review by derrabus — APPROVED** [review]
Looks good. Shall we add a section explaining how to enable a constraint for both usages (Doctrine Annotations + PHP Attributes)? Or is this a problem, an application usually does not have?

File: validation/custom_constraint.rst [hunk 4]
```diff
@@ -12,20 +12,43 @@ alphanumeric characters.
 Creating the Constraint Class
 -----------------------------
 
-First you need to create a Constraint class and extend :class:`Symfony\\Component\\Validator\\Constraint`::
+First you need to create a Constraint class and extend :class:`Symfony\\Component\\Validator\\Constraint`:
 
-    // src/Validator/ContainsAlphanumeric.php
-    namespace App\Validator;
+.. versionadded:: 5.2
```
Comment (OskarStark) [inline validation/custom_constraint.rst]: This should be moved below the configuration block, thanks

Comment (yoannrenard) [reply validation/custom_constraint.rst]: Done @OskarStark

**Review by OskarStark — APPROVED** [review]

**Comment by yoannrenard** [comment]
@derrabus You mean something explaining how to use both (Doctrine annotations + PHP attributes) at the same time? Like using a constraint that supports attributes and one which doesn't?

**Comment by derrabus** [comment]
You can build classes that are a valid Doctrine annotation as well as a valid PHP attribute. Almost all of Symfony's built-in constraints work that way. The examples you provided only work with one of the annotation implementations. But that's probably knowledge most developers won't need.

I think, we're good here.

**Review by derrabus — APPROVED** [review]

File: validation/custom_constraint.rst [hunk 5]
```diff
@@ -12,20 +12,43 @@ alphanumeric characters.
...
+
+        #[\Attribute]
+        class ContainsAlphanumeric extends Constraint
+        {
+            public $message = 'The string "{{ string }}" contains an illegal character: it can only contain letters or numbers.';
+        }
+
+.. versionadded:: 5.2
+
+    The ability to use PHP attributes to configure constraint was introduced in
```
Comment (xabbuh) [inline validation/custom_constraint.rst]: [...] to configure constraint**s** [...]

Comment (yoannrenard) [reply validation/custom_constraint.rst]: fixed @xabbuh

**Review by OskarStark — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by OskarStark** [comment]
Thank you Yoann.

**Pull request merged**